packages/yarnpkg-pnpify/sources/Hoister.ts

```diff
@@ -0,0 +1,289 @@
+        lastPkgId++;
+        const packagePath = npath.toPortablePath(pkg.packageLocation);
+        let weight = 1;
+        if (this.options.optimizeSizeOnDisk && packagePath.indexOf(`.zip${NODE_MODULES_SUFFIX}`) > 0)
+          weight = fs.statSync(packagePath.split(NODE_MODULES_SUFFIX)[0]).size;
+
+        locatorToPackageMap.set(locatorKey, pkgId);
+        locators.push(locator);
+        packages.push({name: locator.name!, weight});
+        packageInfos.push(pkg);
+      }
+      return pkgId;
+    };
+
+    const getPackageId = (locator: PackageLocator) => locatorToPackageMap.get(getLocatorKey(locator));
+
+    const addedIds = new Set<PackageId>();
+    const addPackageToTree = (pkg: PackageInformation<NativePath>, pkgId: PackageId, parentDepIds: Set<PackageId>) => {
+      addedIds.add(pkgId);
+      const deps = new Set<PackageId>();
+      const peerDeps = new Set<PackageId>();
+      packageTree[pkgId] = {deps, peerDeps};
+
+      for (const [name, reference] of pkg.packageDependencies) {
+        if (reference) {
+          const locator = typeof reference === 'string' ? {name, reference} : {name: reference[0], reference: reference[1]};
+          const depPkgId = getPackageId(locator);
+          if (depPkgId) {
+            if (pkg.packagePeers) {
+              if (pkg.packagePeers.has(name)) {
+                peerDeps.add(depPkgId);
+              } else {
+                deps.add(depPkgId);
+              }
+            } else {
+              if (parentDepIds.has(depPkgId) && depPkgId !== pkgId) {
+                peerDeps.add(depPkgId);
+              } else {
+                deps.add(depPkgId);
+              }
+            }
+          }
+        }
+      }
+
+      const allDepIds = new Set(Array.from(deps).concat(Array.from(peerDeps)));
+      for (const depId of allDepIds) {
+        const depPkg = packageInfos[depId];
+        if (depPkg && !addedIds.has(depId)) {
+          addPackageToTree(depPkg, depId, allDepIds);
+        }
+      }
+    };
+
+    const assignedIds = new Set<PackageId>();
+    const assignPackageIds = (pkgId: PackageId) => {
+      assignedIds.add(pkgId);
+      const pkg = packageInfos[pkgId];
+      for (const [name, reference] of pkg.packageDependencies) {
+        if (reference) {
+          const locator = typeof reference === 'string' ? {name, reference} : {name: reference[0], reference: reference[1]};
+          const depPkg = pnp.getPackageInformation(locator);
+          if (depPkg) {
+            const depPkgId = assignPackageId(locator, depPkg);
+            if (!assignedIds.has(depPkgId)) {
+              assignPackageIds(depPkgId);
+            }
+          }
+        }
+      }
+    };
+
+    const pkg = pnp.getPackageInformation(pnp.topLevel)!;
+    const topLocatorKey = getLocatorKey(pnp.topLevel);
+    for (const locator of pnpRoots) {
+      if (getLocatorKey(locator) !== topLocatorKey) {
+        pkg.packageDependencies.set(locator.name!, locator.reference);
+      }
+    }
```

Comment:
That's worth a comment to explain. If I understand it's to put all workspaces dependencies into the root node_modules, correct? What happens when you have two workspaces depending on different versions of a same package?

Reply:
I guess, its the same situation as with v1 and packages/foo and packages/bar depending on different versions of a same package. Do you think we should hoist each pnp tree root separately?